tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -895,6 +898,203 @@ static void domain_remove_part_context_switch(void)
+	obj_end = (char *)stack_obj + obj_size;
+	obj_start = (char *)stack_obj;
+
+	/* Assert that the created stack object, with the reserved data
+	 * removed, can hold a thread buffer of STEST_STACKSIZE
+	 */
+	zassert_true(STEST_STACKSIZE <= (obj_size - K_THREAD_STACK_RESERVED),
+		      "bad stack size in object");
+
+	/* Check that the stack info in the thread marks a region
+	 * completely contained within the stack object
+	 */
+	zassert_true(stack_end <= obj_end,
+		     "stack size in thread struct out of bounds (overflow)");
+	zassert_true(stack_start >= obj_start,
+		     "stack size in thread struct out of bounds (underflow)");
+
+	/* Check that the entire stack buffer is read/writable */
+	printk("   - check read/write to stack buffer\n");
+
+	/* Address of this stack variable is guaranteed to part of
+	 * the active stack, and close to the actual stack pointer.
+	 * Some CPUs have hardware stack overflow detection which
+	 * faults on memory access within the stack buffer but below
+	 * the stack pointer.
+	 *
+	 * First test does direct read & write starting at the estimated
+	 * stack pointer up to the highest addresses in the buffer
+	 */
+	stack_ptr = &val;
+	for (pos = stack_ptr; pos < stack_end; pos++) {
+		/* pos is volatile so this doesn't get optimized out */
+		val = *pos;
+		*pos = val;
+	}
+
+	if (z_arch_is_user_context()) {
+		/* If we're in user mode, check every byte in the stack buffer
+		 * to ensure that the thread has permissions on it.
+		 */
+		for (pos = stack_start; pos < stack_end; pos++) {
+			zassert_false(check_perms((void *)pos, 1, 0),
+				      "bad MPU/MMU permission on stack buffer at address %p",
+				      pos);
+		}
+
+		/* Bounds check the user accessible area, it shouldn't extend
+		 * before or after the stack. Because of memory protection HW
+		 * alignment constraints, we test the end of the stack object
+		 * and not the buffer.
+		 */
+		zassert_true(check_perms(obj_start - 1, 1, 0),
+			     "user mode access to memory before start of stack object");
+		zassert_true(check_perms(obj_end, 1, 0),
+			     "user mode access past end of stack object");
+	}
+
+
+	/* This API is being removed just whine about it for now */
+	if (K_THREAD_STACK_BUFFER(stack_obj) != stack_start) {
+		printk("WARNING: K_THREAD_STACK_BUFFER() reports %p\n",
+		       K_THREAD_STACK_BUFFER(stack_obj));
+	}
+
+	if (z_arch_is_user_context()) {
+		zassert_true(stack_size <= obj_size - K_THREAD_STACK_RESERVED,
+			      "bad stack size in thread struct");
+	}
+
+
+	k_sem_give(&uthread_end_sem);
+}
+
+void stest_thread_entry(void *p1, void *p2, void *p3)
+{
+	bool drop = (bool)p3;
+
+	if (drop) {
+		k_thread_user_mode_enter(stest_thread_entry, p1, p2,
+					 (void *)false);
+	} else {
+		stack_buffer_scenarios((k_thread_stack_t *)p1, (size_t)p2);
+	}
+}
+
+void stest_thread_launch(void *stack_obj, size_t obj_size, u32_t flags,
+			 bool drop)
+{
+	k_thread_create(&uthread_thread, stack_obj, STEST_STACKSIZE,
```

Comment:
Can we repeat the test using k_thread_create with THREAD_STACK_SIZEOF, to confirm that _that_ is also working?

Reply:
can we do this some other time?
I have very little bandwidth to further enhance this test until after 1.14 is out

Reply:
Let's improve this in 1.15